frontend/controller/router.js

```diff
@@ -70,113 +68,133 @@ function createEnterGuards (...guards) {
-      beforeEnter: createEnterGuards(loginGuard, groupGuard)
-    },
-    {
-      path: '/payments',
-      component: lazyPayments,
-      name: 'Payments',
-      meta: { title: L('Payments') },
-      beforeEnter: createEnterGuards(loginGuard, groupGuard)
-    },
-    /* Guards need to be created for any route that should not be directly accessed by url */
-    {
-      path: '/messages',
-      component: lazyMessages,
-      name: 'Messages',
-      meta: { title: L('Messages') },
-      beforeEnter: createEnterGuards(loginGuard)
-    },
-    {
-      path: '/messages/:chatName',
-      component: lazyMessages,
-      name: 'MessagesConversation',
-      beforeEnter: createEnterGuards(loginGuard)
-      // BUG/REVIEW "CANNOT GET /:username" when username has "." in it
-      // ex: messages/joe.kim doesnt work but messages/joekim works fine.
-      // Possible Solution: https://router.vuejs.org/guide/essentials/history-mode.html#example-server-configurations
-    },
-    {
-      path: '/group-chat',
-      component: lazyGroupChat,
-      name: 'GroupChat',
-      meta: { title: L('Group Chat') },
-      beforeEnter: createEnterGuards(loginGuard, groupGuard)
-    },
-    {
-      path: '/group-settings',
-      component: lazyGroupSettings,
-      name: 'GroupSettings',
-      meta: { title: L('Group Settings') },
-      beforeEnter: createEnterGuards(loginGuard, groupGuard)
-    },
-    {
-      path: '/group-chat/:chatRoomId',
-      component: lazyGroupChat,
-      name: 'GroupChatConversation',
-      beforeEnter: createEnterGuards(loginGuard, groupGuard)
-    },
-    {
-      path: '/join',
-      name: Join.name,
-      component: Join,
-      meta: { title: L('Join a Group') },
-      // beforeEnter: createEnterGuards(loginGuard, mailGuard)
-      beforeEnter: createEnterGuards(inviteGuard)
-    },
-    ...(process.env.NODE_ENV === 'development'
-      ? [{
-          path: '/error-testing',
-          name: 'ErrorTesting',
-          component: () => import('../views/pages/ErrorTesting.vue'),
-          meta: { title: L('Error Testing') }
-        }]
-      : []),
-    {
-      path: '*',
-      redirect: '/'
-    }
-  ]
+  }
 })
 
 router.beforeEach((to, from, next) => {
   document.title = to.meta.title
   next()
 })
 
+let startRoute
+router.beforeResolve((to, from, next) => {
+  if (!startRoute) startRoute = to
+  // console.warn(`beforeResolve to: ${to.path} from: ${from.path}`)
+  next()
+})
+
+// Sometimes beforeEnter will get triggered too early b/c of fix for #1192
+// and store.state isn't loaded yet, so we run it again after VUE_LOADED
+sbp('okTurtles.events/once', VUE_LOADED, function () {
+  const route = routes.find(r => r.path === startRoute.path)
+  // $FlowFixMe
+  const next = route?.beforeEnter?.(startRoute, null, (next) => next)
+  // console.warn(`started at ${startRoute.path} and got next of ${JSON.stringify(next)}`)
+  if (next) {
+    router.push(next).catch(logExceptNavigationDuplicated)
+  }
+})
+
```

Comment:
This seems a bit hacky and I have some trouble picturing the control flow here.

Reply:
In fact I just don't understand why all this was necessary. :sweat_smile:

Reply:
It is a bit hacky.

It's necessary because of what's mentioned in the comment. In some cases, some pages will not redirect properly because `store.state` isn't initialized but the guards get triggered early on anyway

Reply:
To elaborate a bit more, `store.state.currentGroupId` will be `null` because the contracts haven't yet finished being loaded by `login`. We delay `login` because it performs a contract sync which would result in a blank white page if that took too long.

Reply:
@snowteamer Actually, I've been thinking about this, and I think there is a cleaner way to implement it, so thanks for pointing this out.

I will work on separating out the contract sync part of the login from the rest of the login action, and that should fix it.

Reply:
@greg Thank you for the additional details.
> I will work on separating out the contract sync part of the login from the rest of the login action, and that should fix it.
Sounds promising, I hope this works :+1:

Reply:
Seems to work. 🙂